src/file.rs

```diff
@@ -0,0 +1,180 @@
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+/// Utilites for db file.
+
+use std::sync::atomic::{AtomicU64, AtomicBool, Ordering};
+use parking_lot::{RwLockUpgradableReadGuard, RwLock};
+use crate::error::Result;
+use crate::table::TableId;
+
+#[cfg(target_os = "linux")]
+fn disable_read_ahead(file: &std::fs::File) -> Result<()> {
+    use std::os::unix::io::AsRawFd;
+    let err = unsafe { libc::posix_fadvise(file.as_raw_fd(), 0, 0, libc::POSIX_FADV_RANDOM) };
+    if err != 0 {
+        Err(std::io::Error::from_raw_os_error(err))?
+    } else {
+        Ok(())
+    }
+}
+
+#[cfg(target_os = "macos")]
+fn disable_read_ahead(file: &std::fs::File) -> Result<()> {
+	use std::os::unix::io::AsRawFd;
+	if unsafe { libc::fcntl(file.as_raw_fd(), libc::F_RDAHEAD, 0) } != 0 {
+		Err(std::io::Error::last_os_error())?
+	} else {
+		Ok(())
+	}
+}
+
+#[cfg(not(any(target_os = "macos", target_os = "linux")))]
+fn disable_read_ahead(_file: &std::fs::File) -> Result<()> {
+	Ok(())
+}
+
+// `File::sync_data` uses F_FULLSYNC fcntl on MacOS. It it supposed to be
+// the safest way to make sure data is fully persisted. However starting from
+// MacOS 11.0 it severely degrades parallel write performance, even when writing to
+// other files. Regular `fsync` is good enough for our use case.
+// SSDs used in modern macs seem to be able to flush data even on unexpected power loss.
+// We performed some testing with power shutdowns and kernel panics on both mac hardware
+// and VMs and in all cases `fsync` was enough to prevent data corruption.
+#[cfg(target_os = "macos")]
+fn fsync(file: &std::fs::File) -> Result<()> {
+	use std::os::unix::io::AsRawFd;
+	if unsafe { libc::fsync(file.as_raw_fd()) } != 0 {
+		Err(std::io::Error::last_os_error())?
+	} else {
+		Ok(())
+	}
+}
+
+#[cfg(not(target_os = "macos"))]
+fn fsync(file: &std::fs::File) -> Result<()> {
+	file.sync_data()?;
+	Ok(())
+}
+
+
+const GROW_SIZE: u64 = 256 * 1024;
+
+pub struct DbFile {
+	pub file: RwLock<Option<std::fs::File>>,
+	pub path: std::path::PathBuf,
+	pub capacity: AtomicU64,
+	pub dirty: AtomicBool,
+	pub origin: TableId,
+}
+
+impl DbFile {
+	pub fn open(filepath: std::path::PathBuf, header_size: u8, entry_size: u16, origin: TableId) -> Result<Self> {
+		let mut capacity = header_size as u64;
+		let file = if std::fs::metadata(&filepath).is_ok() {
+			let file = std::fs::OpenOptions::new().create(true).read(true).write(true).open(filepath.as_path())?;
+			disable_read_ahead(&file)?;
+			if file.metadata()?.len() == 0 {
+				// Preallocate.
+				debug_assert!(header_size as u64 * entry_size as u64 <= GROW_SIZE);
+				file.set_len(GROW_SIZE)?;
```

Comment:
The file size should not be `GROW_SIZE` exactly, but rather `capacity` * `entry_size`. 
`GROW_SIZE` may contain a non-integer number of records.